Enable test and fix a bug in directcall

rwy7 opened the pull request.

rwy7 committed:
Remove any remaining b9main functions

Stop mentioning b9main in the docs. The function is now called body.

From the tests, move the body of b9main to the body of the script.

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Make stack printer sexpr based

It's a little easier to tell when the stack printout starts and stops.

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Add ValueBuilder and add explicit push/popInt48 calls

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Add ValueBuilder value helpers

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Add trace and print stack helpers for the jit

Helpful for debugging

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Trace compiled method execution when debug is on

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Remove unused variable

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy7 committed:
Call interpret_0 when not using passparam

When pass param is disabled, we are never passing any parameters in
registers. Always call interpret_0, regardless of the number of
parameters. The parameters will be passed on the operand stack.

Signed-off-by: Robert Young <rwy0717@gmail.com>

youngar commented on docs/build-a-runtime/index.md:
```diff
@@ -89,12 +89,10 @@ The above diagram details the two high-level components of the base9 architectur
 Our frontend language is a primitive subset of JavaScript. Currently, it can only operate on integers. Let's have a look at some code:
```
Lol we still need to fix this, feel like tacking it on?

rwy7 replied:
haha yeah... I have it fixed in a different branch somewhere, which basically rewrites the intro. I'm trying to avoid big changes to the docs until that goes in. I can probably get to it later today though.

youngar replied:
Ahh I see

youngar commented on docs/build-a-runtime/index.md:
````diff
@@ -89,12 +89,10 @@ The above diagram details the two high-level components of the base9 architectur
 ```
 
-Above is a classic program that we all know and love, Hello, World! The source code can be found in [test/hello.src]. The `b9main` function is always the entry point to a program, like `int main` in C. `b9PrintString` is a method from a tiny library of functions that will be compiled with every program. The library is called b9stdlib and can be found in [b9/js_compiler/b9stdlib.src]. The function ...
+ `b9PrintString` is a method from a tiny library of functions that will be compiled with every program. The library is called b9stdlib and can be found in [b9/js_compiler/b9stdlib.src]. The functions in b9stdlib all call base9 [primitive functions]. In the Hello, World! program, Our script uses `b9PrintString` from b9stdlib to write text to console.
````
It is probably not accurate to say all b9stdlib functions call primitive functions.. even though right now they do.  I think in the future we're going to change how primitive functions are implemented, so we remove our weird cffi entirely.  For now I guess its fine, just a thought.

youngar commented on docs/build-a-runtime/index.md:
```diff
@@ -209,22 +207,20 @@ Let's run the disassembler using the binary module we generated in the [Frontend
-(function "b9main" 0 0
+(function "<body>" 0 0
   0  (str_push_constant 0)
   1  (function_call 0)
```
Have we thought about a good way to get the dissembler to print other information (like function names?).  We could just add a `--verbose` that adds it in lispy style comments.

rwy7 replied:
I've been thinking about this. Right now, our printer just relies on `std::ostream` streaming of individual values. Better printouts are going to need more context, probably the complete module.

I'm considering introducing a second set of APIs where we can print values with contextual information:

* `printModule(out, m)`
* `printFunction(out, m, f)`
* `printInstruction(out, m, i)`

That, and possibly contributing the sexpr printer from ab to b9, for better formatting.

rwy7 committed:
Only call factorial up to ten

Previously, we were calling factorial on 1 to 20. factorial(20) results
in an integer that is too large to be represented in 48bits.

To support factorial(20), we have to implement promotion of integers to
doubles in the VM. We do not yet support doubles at all.

This calculation was underflowing in the interpreter, and causing badly
tagged values in the JIT.

Fixes the factorial test in CI.

Signed-off-by: Robert Young <rwy0717@gmail.com>

youngar commented:
@rwy0717 LGTM, are we good to merge?

rwy7 commented:
Yeah, I think so.
